### DISCUSSION TITLE
Pie Chart with Labels and facet in altair: not able to proper render the value labels

### DISCUSSION THREAD

**Discussion by aborruso** [vega/altair#2887] [Q&A] [answered]
Using the [single chart example][1], I have a label for each piece of cake.

![image](https://user-images.githubusercontent.com/30607/218721513-3efc61f3-de0d-4be0-9ed4-edc83e6bc2e6.png)

If I try to transform it in a faceted chart using this code

```python
df=pd.read_csv("input.csv",keep_default_na=False)

base=alt.Chart(df).encode(
    theta=alt.Theta(field="v", type="quantitative"),
    color=alt.Color(field="k", type="nominal")
)


pie = base.mark_arc(outerRadius=100)
text = base.mark_text(radius=115,fill= "black").encode(alt.Text(field="v", type="quantitative", format=",.1f"))

alt.layer(pie, text, data=df).facet(column='label')
```

all the labels are all in the same wedge and then illegible (here the vega lite version [vega lite version](https://vega.github.io/editor/#/url/vega-lite/N4Igxg9gdgZglgcxALlANzgUwO4tJKAFzigFcJSBnAdTgBNCALFAZgAY2AacaYsiygAlMiRoVYcAvpO50AhoTl4QUOQFtMKEPMUBaAEYsA7AA5MARjYBWfZkxWjAFgCczjibkAmFleefn5gEudCAyIDByYJjiqDwANqRqUMrwmHEhyCBxcrZxINyEAJ4ADpqZUBBqJHJ50tyUpWDK2YWYAE4oANqganJtANbKRaVafU3cFITtAEpydHBUKJZsYZhQkPNQSLGQcRAdsanpWoMFJWUqldW1BYzRSodYx5lo+SDDFwCOpHJEcIrENCaaQyHp9QaxD5aKYAD3E3HgcTymX02TApxAbTmC0oS3MVlW6wgm22+AgewOoCOGRAGKh5Suqhu70wcJSTxprwR+16MXyADpzDA3vSQN9fsQAXAgaFbvd2WlOSLzlpxX8pTKQQBdMIAEkoYDuvS0YkIxUoyAA9JagQg5PyEP9GKR9Py4BBLQajXIbZg7bo4v9MDarPzPPy2PyAFaUaBvHRySjRXGxBMGYxmSw2OwOFxuNgeby+fyBczBLpUpbcZLIFjcCEgNi6KwAAgApG9XshXPyjJ4TK4+wPnH3nNxsrktAAxfaYSjiUHhKsqFCOetaKy6SztzsoSOeIzLA9Hw8cTzjnJpaez+eyyvILgr5DmdeZJutjvcLvLC+TzIAOWgXQZzaOcF04e9PCsNcnysV9G2bHcvxQHsOCMQ90KsAtPAANjHLJL2REBSHWbJKEoOB4ByOJgQgpdkCgmCaxw+DN23T8QC7SNXE8fteJccwjHYZwTHPAi-2I0jEwoqjUWBHVJCAA)).

![image](https://user-images.githubusercontent.com/30607/218721575-ef845e22-4a55-4d94-86d3-69e76e4a048b.png)

How to have a result like that of the single chart?

Thank you

```
f,n,k,v,label
1,3,0-5 %,99.7289972899729,Forest
1,4,5-10 %,0.27100271002710025,Forest
0,1,0-5 %,100.0,Non-Forest
254,5,0-5 %,99.0077177508269,unclassifiable
254,6,5-10 %,0.9922822491730982,unclassifiable
```

  [1]: https://altair-viz.github.io/gallery/pie_chart.html
  [2]: https://i.stack.imgur.com/f6Tym.png
  [3]: https://i.stack.imgur.com/ftR97.png

**Comment by ChristopherDavisUCI** [comment]
Does adding the following to the very end make it work the way you want?
```
.resolve_scale(
    theta="independent"
)
```
By default, Vega-Lite and Altair will try to use a shared scale for all three of the pie charts.

Here is an example that I tried.  (You can also remove the `resolve_scale` to see what impact it's having.)

```
import pandas as pd
import altair as alt

source = pd.DataFrame({"category": [1, 2, 3, 4, 5, 6], 
    "value": [4, 6, 10, 3, 7, 8], 
    "label": ["A", "A", "B", "B", "B", "A"]})

alt.Chart(source).mark_arc().encode(
    theta=alt.Theta(field="value", type="quantitative"),
    color=alt.Color(field="category", type="nominal"),
).facet(
    column="label"
).resolve_scale(
    theta="independent"
)
```

[Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4Igxg9gdgZglgcxALlANzgUwO4tJKAFzigFcJSBnAdTgBNCALFAFgAY2AacaYsiygAlMiRoRQBmDgF9p3OgENCCvCCgKAtphQhFygLQB2BRLosAbAA4ATCwCsYTHUtg6142AkAjc9buXfQ3MYLxA5EBgFR3FUHgAbUg0oVXhMOLodOIUvNJBuQgBPAAdtZDUIDRIFOLDwyhKwVQ0FACcAax1Wxu5MKEg6EiRYyDiIFpSsdJ0wJUwEMYK8kEKSnSgKqprwpkxlCbSMsrRq0m184tKQAEdSBSI4ZWI0bVlwlsxKCDjn1UoZuNKoB2ezKJDomBKUHBRFq3AAJH9GJhmjoxIQipRkAB6LHPBAKAB0CAejFIXgJcAgWMRyIUuLmCn0cQemFxLAJAEZDAS2ASAFafZLyJQKSi7SiqPSM4ymCw2eyOZyudxRby+fyBYKhZAAbXws3mLUWyA53GOCVKLG4WRyNTKAEEwpx9YQ5gsUNYzSdSuZrdlcg6nS63UbJF6LSgOVwQDaAyAAEJB8AG93IK0gc2nMMx-12hNJmauw3GuzhrPIQx+206RNyYPFlC+jPelCWKtxx3SAC6siAA)

**Reply by aborruso** [reply]
> Does adding the following to the very end make it work the way you want?

Hi, thank you.

It remains the same ([link](https://vega.github.io/editor/#/url/vega-lite/N4Igxg9gdgZglgcxALlANzgUwO4tJKAFzigFcJSBnAdTgBNCALFAZgAY2AacaYsiygAlMiRoVYcAvpO50AhoTl4QUOQFtMKEPMUBaAEYsA7AA5MARjYBWfZkxWjAFgCczjibkAmFleefn5gEudCAyIDByYJjiqDwANqRqUMrwmHEhyCBxcrZxINyEAJ4ADpqZUBBqJHJ50tyUpWDK2YWYAE4oANqganJtANbKRaVafU3cFITtAEpydHBUKJZsYZhQkPNQSLGQcRAdsanpWoMFJWUqldW1BYzRSodYx5lo+SDDFwCOpHJEcIrENCaaQyHp9QaxD5aKYAD3E3HgcTymX02TApxAbTmC0oS3MVlW6wgm22+AgewOoCOGRAGKh5Suqhu70wcJSTxprwR+16MXyADpzDA3vSQN9fsQAXAgaFbvd2WlOSLzlpxX8pTKQQBdMJtTCUcky2KUMA1MqgJjyzIkOiYUpQW1EUJhAAkJruvS0YkIxUoyAA9P6gQg5PyEP9GKR9Py4BB-e7ML0g5gQ7o4v9MEGrPzPPy2PyAFYG5KyBRySjRXGxHRyAzGMyWGx2BwuNxsDzeXz+QLmYJdKlLbjJZAsbgQkBsXRWAAEAFI3q9kK5+UZPCZXKv185V85uNlcloAGL7fXiUHhQcqFCOMdaKy6SxzhcofOeIzLN8f98cTx7nJpI8T0oM9OAHZAuCvZBzFvTJJxneduEXZY-wPTIADloF0Y89WA2UwM8Kwb0gqwYInKcn0QlBlw4Ix31oqx208AA2Xcsn-ZEQFIdZskoSg4HgHI4mBUCL2QAiiOHJjSPvR8EJARd81cTw12UlxzCMdhnBMX82NQzjuPLPiBNRYEdUkIA))

**Reply by ChristopherDavisUCI** [reply]
Hmm, I'm not sure.  I tried another example using numbers that differed more greatly (like in your example), and it still looked okay.  Do you see what's different between my example and yours?

<img width="753" alt="Screenshot 2023-02-14 at 5 45 24 AM" src="https://user-images.githubusercontent.com/82750240/218756304-d120687f-1872-4170-9454-d100f5e9869b.png">

```
import pandas as pd
import altair as alt

source = pd.DataFrame({"category": [1, 2, 3, 4, 5, 6], 
    "value": [0.14, 0.16, 10, 3, 7, 8], 
    "label": ["A", "A", "B", "B", "B", "A"]})

base = alt.Chart(source).encode(
    theta=alt.Theta("value:Q", stack=True),
    color=alt.Color("category:N", legend=None),
)

pie = base.mark_arc(outerRadius=120)
text = base.mark_text(radius=140, size=20).encode(
    text="category:N"
)

(pie + text).facet(
    column="label"
).resolve_scale(
    theta="independent"
)
```

[Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4Igxg9gdgZglgcxALlANzgUwO4tJKAFzigFcJSBnAdTgBNCALFAFgAY2AacaYsiygAlMiRoRQBmDgF9p3OgENCCvCCgKAtphQhFygLQB2QwEYwmQwA4AnACM2MAGwsY1mG3swwMBY7DW6SwAmGDo6EDkQH3NxVB4AG1INKFV4THjw5BB4hVt0kG5CAE8AB20sqAgNEgV4iMjKMrBVHKLMACcUAG1QDQV2gGtVYrKdfubuCkIOgCUFOjgqFBMgtkjMKEgFqCQ4yHiITri0jJ0wJUwEQ6KC7MuNzLJ4+MLS8rUqmrrIpkxlVKwpyyaFqpG03EoyjAQ2QhHaYNeoyyAEdSAoiHBlMQ0NpZHJev0YaARu9pgAPcTcdrzRaUZbsCFwABe5VW602EG2u3wEAOR1AJ0y4AuV3aN248XuUEepGeiPelWq6m+hUwFIB6SF52movFIBJOkVXwihUYfxUx0BQpBiXBIEhCmhKDhCP1bx0qPRxCxcBx9WkAF1Iu1MJReX64pRzpLhmb-lkSHRMGVpRtxLJuAASKNmvo6MSEEqUZAAehLOIQCgAdAhMYxSLYq3AICWc5g+uXLgp9PFMZhyywqyZDFW2FWAFZhlLyJQKSh-OlxPTd4xmCw2TzOVzuTzeXz+QIhMLdfAi67Lbg2sEoMcmFgS3L5LIAQRNp515+QQUvoPKt8cD55HUL5vsKH5ipIP62ssXDZI+wEgAAQqB2qXJ+94gFe5QSIBT5IShZ4QcgACsUHXsghi4QhyH4mBaFEQBmG-iglhUTor6BrIQA)

**Reply by mattijn** [reply]
Difference seems to be in the `stack=True`  as channel encoding option in the `theta` channel encoding.

From here: https://vega.github.io/vega-lite/docs/arc.html#pie-and-donut-charts

> Note: For now, [you need to add `stack=True`](https://github.com/vega/vega-lite/issues/5078) to theta to force the text to apply the same polar stacking layout.

So it becomes:

```python
base = alt.Chart(s, width=10, height=10).encode(
    theta=alt.Theta("v:Q", stack=True),
    color=alt.Color("k:N"),
)

pie = base.mark_arc(outerRadius=30)
text = base.mark_text(radius=40, size=10).encode(
    text="k:N"
)

(pie + text).facet(
    column="label"
).resolve_scale(
    theta="independent"
)
```
![image](https://user-images.githubusercontent.com/5186265/218765336-aa4d8672-bd29-4026-ad21-fa56557b9422.png)

**Reply by ChristopherDavisUCI** [reply]
Thanks @mattijn! How did you create `s`?  Did you save the data in an external csv file, or is there a more direct way?

**Reply by aborruso** [reply]
> Difference seems to be in the `stack=True` as channel encoding option in the `theta` channel encoding.

Yes, this [works](https://vega.github.io/editor/#/url/vega-lite/N4Igxg9gdgZglgcxALlANzgUwO4tJKAFzigFcJSBnAdTgBNCALFAZgAY2AacaYsiygAlMiRoVYcAvpO50AhoTl4QUOQFtMKEPMUBaAEYsA7AA5MARjYBWfZkxWjAFgCczjibkAmFleefn5gEudCAyIDByYJjiqDwANqRqUMrwmHEhyCBxcrZxINyEAJ4ADpqZUBBqJHJ50tyUpWDK2YWYAE4oANqganJtANbKRaVafU3cFITtAEpydHBUKJZsYZhQkPNQSLGQcRAdsanpWoMFJWUqldW1BYzRSodYx5lo+SCUimCDyIRtpJhnEaZACOpDkRDgimIaE00hkPT631AwwuUwAHuJuPA4nlMvpsl83m05gtKEtzFZVusIJttvgIHsDqAjhkQKcQCitBUqqobhzMBiUk9Wa8sftejF8gA6cwwN6ckFgiFQuAw0K3e5CtIit4fSLfX7-QEXUHg4gqtVwgC6YTamEoDLVsUoYBqZWRd0UWhIdEwpSgvqIoTCABIXXdeloxIRipRkAB6eMwhByKUISGMUj6KVwCDx8OYXpJzAp3RxSGYJNWKWeKVsKUAKwdyVkCjklGiZNiOjkBmMZksNjsDhcbjYHm8vn8gXMwS6zKW3GSyBY3G+IDYuisAAIAKRvV7IVxSoyeEyuU-n5yn5zcbK5LQAMX29vE8PCi5UKEca60Vl0lh7geKD1p4RjLGBEHgRwnh3jkaRPi+HzqguyBcF+yDmL+mSbju+7cIeyxwQ+mQAHLQLoz52sh75ysgnhWD+GFWNhG5bkBBEoMeHBGOBvFWOOngAGy3lk8G4iApDrNklCUHA8A5HEsKcKhDFMcuQmsf+gH4SAh71q4nhnkZLjmEY7DOCYsFiSRknSe2ckKfisI2pIQA), thank you very much!!!

**Reply by mattijn** [reply]
Quickly copied it from the vega-editor url:
```
s = pd.DataFrame.from_records([
      {"f": 1, "n": 3, "k": "0-5 %", "v": 99.7289972899729, "label": "Forest"},
      {
        "f": 1,
        "n": 4,
        "k": "5-10 %",
        "v": 0.2710027100271002,
        "label": "Forest"
      },
      {"f": 0, "n": 1, "k": "0-5 %", "v": 100, "label": "Non-Forest"},
      {
        "f": 254,
        "n": 5,
        "k": "0-5 %",
        "v": 99.0077177508269,
        "label": "unclassifiable"
      },
      {
        "f": 254,
        "n": 6,
        "k": "5-10 %",
        "v": 0.9922822491730982,
        "label": "unclassifiable"
      }
    ])
```

**Comment by aborruso** [comment, marked as answer]
I'm adding here the link to @mattijn to mark this as the right answer:
https://github.com/altair-viz/altair/discussions/2887#discussioncomment-4971355
